File: SlidingWindow/ContainsDuplicate2.java
package SlidingWindow;
import java.util.HashSet;
import java.util.Set;

public class ContainsDuplicate2 {
    
    public static boolean containsDuplicate(int[] nums, int k) {
        Set<Integer> window = new HashSet<>();
        int left = 0;

        for (int right = 0; right < nums.length; right++) {
            if (right - left > k) {
                window.remove(nums[left]);
                left++;
            }

            if (window.contains(nums[right])) {
                return true;
            }

            window.add(nums[right]);
        }

        return false;
    }

    public static void main(String[] args) {
        int[] arr = {1,2,3,1};
        int k = 3;
        boolean res = containsDuplicate(arr, k);
        System.out.println(res);
    }
}

File: SlidingWindow/ContainsDuplicate2.py
from typing import List
class ContainsDuplicate2:
    def containsDuplicate(self, nums: List[int], k: int) -> bool:
        nums_set = set()
        left = 0

        for right in range(len(nums)):
            if right - left > k:
                nums_set.remove(nums[left])
                left += 1

            if nums[right] in nums_set:
                return True

            nums_set.add(nums[right])

        return False

def main():
    sol = ContainsDuplicate2()
    nums = [1,2,3,1,2,3]
    k = 2
    res = sol.containsDuplicate(nums, k)
    print(res)

if __name__ == "__main__":
    main()
    
